### PR TITLE
feat: IP Identity

### LINKED ISSUE #75: Mediolano IP ID Onchain - Cairo Smart Contract
IP ID serves as a identity protocol for Programmable IP in Starknet, facilitating storage and sharing of specific data within the IP ID certified page and Mediolano ecosystem. This IP identity protocol allows various Starknet app to access and utilize user IP information effortlessly, enhancing integrations across the blockchain.

When a user registers a IP ID, this information is accessible across the Starknet Ecosystem. This means that every app will can share and use the IP ID info and Programmable licensing.

Criteria:

[ ] Develop the Cairo Smart Contract.
[ ] Support permissionless minting and purchasing of NFTs for community access.
[ ] Validate ownership and enable access to exclusive features.
[ ] Provide thorough testing and detailed documentation for the smart contract.

Technical Details:
Platform: Starknet Blockchain
Language: Cairo
Dependencies: Integration with Mediolano dApp and Starknet 
Security: Implement mechanisms to prevent unauthorized access and ensure data integrity.
Scalability: Optimize the contract to handle high volumes.

Testing:
Conduct comprehensive testing to validate all functionalities, including community creation, minting, and exclusive access mechanisms.

Documentation:
Create detailed technical documentation and user guides for developers and community creators.

### PR DESCRIPTION
**Pull request by anonfedora**
# Implement IP Identity Cairo Smart Contract for Mediolano IP ID Onchain (#75)

Closes #75

## Description

This PR implements the Mediolano IP ID Onchain Cairo Smart Contract as specified in issue #75. The contract serves as an identity protocol for Programmable IP on Starknet, enabling permissionless minting and purchasing of NFTs, ownership validation, and access to exclusive features within the Mediolano ecosystem. The implementation includes comprehensive testing and detailed documentation to ensure robustness and usability.

### Smart Contract Development:
- Created `IPIdentity.cairo`, a Cairo smart contract that implements the `IIPIdentity` interface with the following functionalities:
  - `register_ip_id`: Allows users to register an IP ID with metadata, IP type, and license terms, minting an ERC721 NFT.
  - `update_ip_id_metadata`: Enables owners to update the metadata URI of their IP ID.
  - `get_ip_id_data`, `get_ip_id_by_owner`, `get_owner_by_ip_id`: Retrieve IP ID details and ownership information.
  - `verify_ip_id`: Allows the contract owner to verify an IP ID.
  - `set_payment_token`, `set_paused`: Administrative functions for managing payment tokens and contract pausing.
- Integrated OpenZeppelin Cairo contracts (v0.17.0) for ERC721, Ownable, Pausable, and SRC5 functionality.
- Implemented payment processing with a configurable ERC20 token (default mint price: 0.001 ETH).
- Added security mechanisms to prevent unauthorized access (e.g., ownership checks, pausable state).
- Optimized storage with mappings for IP ID data and ownership tracking.

### Testing:
- Added `test_contract.cairo` with comprehensive unit tests covering:
  - Successful and failed scenarios for metadata updates, IP ID verification, and administrative functions.
  - Edge cases like invalid IP IDs, non-owner access, and paused contract states.
- Used `snforge_std` for testing with mocked caller addresses and timestamps.
- Tests validate minting, ownership, and exclusive access mechanisms.

### Configuration and Build:
- Added `Scarb.toml` with dependencies: starknet (2.8.4), openzeppelin (v0.17.0), snforge_std (0.36.0), and assert_macros (2.8.4).
- Included `Scarb.lock` for reproducible builds.
- Configured `snfoundry.toml` for testing and deployment settings.
- Added `.gitignore` to exclude build artifacts (target, .snfoundry_cache/).

### Directory Structure:
```
onchain/ip_id/
├── .gitignore
├── Scarb.lock
├── Scarb.toml
├── snfoundry.toml
├── src/
│   ├── IPIdentity.cairo
│   └── lib.cairo
├── tests/
│   └── test_contract.cairo
```

## Criteria Fulfillment

- **Develop the Cairo Smart Contract:** Implemented `IPIdentity.cairo` with all required functionalities.
- **Support permissionless minting and purchasing of NFTs:** `register_ip_id` allows anyone to mint an IP ID NFT with payment.
- **Validate ownership and enable access to exclusive features:** Ownership checks in `update_ip_id_metadata` and `verify_ip_id` ensure only authorized users can modify or verify IP IDs.
- **Provide thorough testing and detailed documentation:**
  - Comprehensive tests in `test_contract.cairo`.
  - Documentation included in code comments and external README.md (to be added in a follow-up commit if required).

## Technical Details

- **Platform:** Starknet Blockchain
- **Language:** Cairo
- **Dependencies:** Integrated with OpenZeppelin for ERC721 and Mediolano dApp compatibility.
- **Security:** 
  - Ownership validation using OpenZeppelin's OwnableComponent.
  - Pausable functionality to mitigate risks.
  - Input validation to prevent invalid IP ID registration or updates.
- **Scalability:** 
  - Efficient storage using Map for IP ID data and ownership.
  - Optimized event emissions for tracking (IPIDRegistered, IPIDUpdated, PaymentTokenUpdated).

## Testing Instructions

1. Navigate to `onchain/ip_id/`.
2. Run tests using:
```bash
scarb test
```
3. Verify test coverage for minting, ownership validation, and error handling.

## Documentation

- Inline comments in `IPIdentity.cairo` explain key functions and logic.
- Tests in `test_contract.cairo` serve as usage examples.
- A detailed README.md will be added in a follow-up commit to include user guides and developer instructions, as per the issue requirements.

## Additional Notes

- The contract assumes a dummy payment token for testing; integration with the actual Mediolano payment token should be configured post-deployment.
- Suggestions for additional features (e.g., batch minting, metadata versioning) can be discussed for future iterations.

## Commits

- **feat: IP Identity** (Apr 14, 2025, by @anonfedora)
  - Added 651 lines across 7 files: .gitignore, Scarb.lock, Scarb.toml, snfoundry.toml, IPIdentity.cairo, lib.cairo, test_contract.cairo.

## Checklist
Screenshot_20250414_133207

- [x] Code follows Starknet and Cairo best practices.
- [x] Tests cover critical functionalities and edge cases.
- [x] Dependencies are correctly configured.
- [x] Contract is secure and scalable.

## Reviewer Instructions

- Verify contract functionality using the test suite.
- Check security mechanisms (e.g., ownership checks, pausable state).
- Review test coverage and suggest additional test cases if needed.
- Confirm compatibility with Mediolano dApp integration requirements.

Please let me know if additional changes or clarifications are needed!

### CODE REVIEW

File: onchain/ip_id/src/IPIdentity.cairo [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        ) -> u256 {
+            self.pausable.assert_not_paused();
+
+            // Check if IP ID is already registered
+            assert(self.ip_id_to_owner.read(ip_id).is_zero(), ERROR_ALREADY_REGISTERED);
+
+            let caller = get_caller_address();
+
+            // Process payment
+            self._process_payment(caller, MINT_PRICE);
```
Comment (PedroRosalba) [inline onchain/ip_id/src/IPIdentity.cairo:185]: this process_payment does not make much sense since the user's already going to have to pay to mint the NFT automatically (it's a blockchain event, the user's wallet is connected)

File: onchain/ip_id/src/IPIdentity.cairo [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        #[substorage(v0)]
+        erc721: ERC721Component::Storage,
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        #[substorage(v0)]
+        pausable: PausableComponent::Storage,
+        #[substorage(v0)]
+        src5: SRC5Component::Storage,
+        ip_id_to_owner: Map<felt252, ContractAddress>,
+        owner_to_ip_id: Map<ContractAddress, felt252>,
```
Comment (PedroRosalba) [inline onchain/ip_id/src/IPIdentity.cairo:114]: do you need both storages ? in my mind that's like "redundant" (they're not exactly the same but serve similar functionalities, so consider brushing this to a simpler resolution

File: onchain/ip_id/src/IPIdentity.cairo [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+            self.emit(IPIDUpdated { ip_id, owner });
+        }
+
+        fn set_payment_token(ref self: ContractState, token_address: ContractAddress) {
+            self.ownable.assert_only_owner();
+            self.mint_payment_token.write(token_address);
+            self.emit(PaymentTokenUpdated { new_token: token_address });
+        }
+
+        fn set_paused(ref self: ContractState, paused: bool) {
```
Comment (PedroRosalba) [inline onchain/ip_id/src/IPIdentity.cairo:267]: do you need the pausable component?

**Review by PedroRosalba — CHANGES_REQUESTED** [review]

**Comment by anonfedora** [comment]
Comments resolved.
Screenshot_20250417_135852
